xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc

```diff
@@ -0,0 +1,1708 @@
+    if (chn_map_.contains(name)) {
+      return absl::InternalError(absl::StrCat("Duplicate channel name ", name));
+    }
+    chn_map_[name] = chn;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<SymbolRefAttr> getChn(std::string name) {
+    auto op = chn_map_.find(name);
+    if (op == chn_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown channel name  ", name));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setMlirValue(int64_t id, Value op) {
+    if (ssa_map_.contains(id)) {
+      return absl::InternalError(
+          absl::StrCat("Duplicate assignment to value id ", id));
+    }
+    ssa_map_[id] = op;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<Value> getMlirValue(int64_t id) {
+    auto op = ssa_map_.find(id);
+    if (op == ssa_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown value id ", id));
+    } else {
+      return op->second;
+    }
+  }
+
+  void newFunctionBaseContext(bool is_proc) {
+    this->ssa_map_ = {};
+    this->is_proc_ = is_proc;
+    this->state_elem_map = {};
+  }
+
+  void setStateElem(std::string name, Value elem) {
+    assert(is_proc_ && "Only proc functions contain state elements");
+    state_elem_map[name] = elem;
+  }
+
+  absl::StatusOr<Value> getStateElem(std::string id) {
+    assert(is_proc_ && "Only proc functions contain state elements");
+    auto elem = state_elem_map.find(id);
+    if (elem == state_elem_map.end()) {
+      return absl::InternalError(absl::StrCat("Unknown state elem ", id));
+    } else {
+      return elem->second;
+    }
+  }
+
+ private:
+  // --- package/module state ----
+  // Maps name -> MLIR ref for all functions in current
+  // package/module.
+  std::unordered_map<std::string, FlatSymbolRefAttr> fn_map_;
+
+  // Maps name -> MLIR ref for all channels in current
+  // package/module.
+  std::unordered_map<std::string, FlatSymbolRefAttr> chn_map_;
+  // Maps XLS source file ID -> pathsourcefile_map_.
+  std::unordered_map<::xls::Fileno, StringAttr> fileno_map_;
+
+  // --- function base (func/proc) state ----
+  // Track if current function base is a proc or plain function. At the moment
+  // this is for assertions/diagnostics only.
+  [[maybe_unused]] bool is_proc_;
+
+  // Maps id -> MLIR value for all SSA values inside the current
+  // function base (function or proc body).
+  std::unordered_map<int64_t, Value> ssa_map_;
+
+  // Maps name -> MLIR value for all state elements of this proc.
+  std::unordered_map<std::string, Value> state_elem_map;
+};
+
+//===----------------------------------------------------------------------===//
+// Type Translation
+//===----------------------------------------------------------------------===//
+
+Type translateType(::xls::Type* xls_type, OpBuilder& builder,
+                   MLIRContext* ctx) {
+  switch (xls_type->kind()) {
+    case ::xls::TypeKind::kTuple: {
+      std::vector<Type> types{};
```

Comment:
{} not needed (any non-POD type has their default constructor called simply by `T t;`, dont need `T t{};`